File: Java/HashMap/CustomLinearProbingHashMap.java
package HashMap;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class CustomLinearProbingHashMap<K,V> implements ICustomHashMap<K,V> {

    private float loadFactor = 0.75f;
    private int size;
    private int capacity;
    private LinearProbingEntry<K,V>[] buckets;

    @Override
    public boolean isEmpty() {
        return this.size == 0;
    }

    @SuppressWarnings("unchecked")
    @Override
    public void initBuckets(int length) {
        this.capacity = length;
        this.buckets =  new LinearProbingEntry[capacity];
    }

    /**
    * Average case O(1), Worst case O(n)
    * Retrieves the value associated with the specified key.
    */
    @Override
    public V get(K key) {
        int bucketIndex = findKey(key);
        if(
            bucketIndex != -1
            && this.buckets[bucketIndex]!= null
            && !this.buckets[bucketIndex].getIsDeleted()
        ) {
            return this.buckets[bucketIndex].getValue();
        }
        return null;
    }

    /**
    * Average case O(1), Worst case O(n)
    * Inserts a key-value pair into the hash table. Resizes the table if necessary.
    */
    @Override
    public void put(K key, V value) {
        int bucketIndex = this.findKey(key);
        if(bucketIndex != -1 
            && !this.buckets[bucketIndex].getIsDeleted()
        ) {
            this.buckets[bucketIndex].setValue(value);
            return;
        }
        int load = (this.size+1)/this.capacity;
        if(load > this.loadFactor) {
            this.resizeHashTable();
        }
        int newIndex = this.findEmptyIndex(key);
        this.buckets[newIndex] = new LinearProbingEntry<K,V>(key, value);
        this.size++;        
    }

    /**
    * O(n)
    * Returns a set of all keys in the hash table.
    */
    @Override
    public Set<K> keySet() {
        Set<K>keys = new HashSet<>();
        for (LinearProbingEntry<K,V> entry : this.buckets) {
            if(entry!= null && !entry.getIsDeleted()) {
                keys.add(entry.getKey());
            }
        }
        return keys;
    }

    /**
    * O(n)
    * Returns a list of all values in the hash table.
    */
    @Override
    public List<V> values() {
        List<V>values = new ArrayList<>();
        for (LinearProbingEntry<K,V> entry : this.buckets) {
            if(entry!= null && !entry.getIsDeleted()) {
                values.add(entry.getValue());
            }
        }
        return values;
    }

    /**
    * O(n)
    * Returns a list of all entries in the hash table.
    */
    @Override
    public List<IEntry<K, V>> entries() {
        List<IEntry<K, V>>values = new ArrayList<>();
        for (LinearProbingEntry<K,V> entry : this.buckets) {
            if(entry!= null && !entry.getIsDeleted()) {
                values.add(entry);
            }
        }
        return values;
    }

    /**
    * Average case O(1), Worst case O(n)
    * Removes the entry with the specified key from the hash table.
    */
    @Override
    public IEntry<K, V> remove(K deleteKey) {
        int bucketIndex = this.findKey(deleteKey);
        if(bucketIndex != -1 ) {
            this.buckets[bucketIndex].setIsDeleted(true);
            this.size--;
            return new LinearProbingEntry<K,V>(buckets[bucketIndex].getKey(),buckets[bucketIndex].getValue());
        }
        return null;
    }

    /**
    * Average case O(1), Worst case O(n)
    * Checks if the hash table contains the specified key.
    */
    @Override
    public boolean containsKey(K key) {
        return this.keySet().contains(key);
    }

    /**
    * O(n)
    * Clears the hash table.
    */
    @Override
    public void clear() {
        Arrays.fill(buckets, null);
        size = 0;
    }

    /**
    * O(n^2)
    * Resizes the hash table and rehashes all non-deleted entries.
    */
    @Override
    public void resizeHashTable() {
        System.out.println("Resizing the Hash Table O(n^2) Operation");
        LinearProbingEntry<K,V>[] oldEntries = this.buckets;
        this.initBuckets(oldEntries.length*2);
        this.size =0;
        for (LinearProbingEntry<K,V> bucketEntry : oldEntries) {
            if(bucketEntry != null && !bucketEntry.getIsDeleted()){
                this.put(bucketEntry.getKey(), bucketEntry.getValue());
            }
        }
    }

    /**
    * O(|key|)
    * Computes the hash of the key.
    */
    @Override
    public int hashFunction(K key) {
        // int hashCode = key.hashCode();
        // return Math.abs(hashCode) % this.capacity;
        int hash = 0;
        String strKey = String.valueOf(key);
        for (int i = 0; i < strKey.length(); i++) {
            char ch = strKey.charAt(i);
            hash = (hash << 5) - hash + ch;
            hash |= 0; // Convert to 32-bit integer
        }
        return hash % buckets.length;
    }

    /**
    * Average case O(1), Worst case O(n)
    * Finds the key in the hash table.
    */
    private int findKey(K key){
        int bucketIndex = this.hashFunction(key);
        if (
            this.buckets[bucketIndex] != null
            && !this.buckets[bucketIndex].getIsDeleted()  
            && this.buckets[bucketIndex].getKey() == key
        ) {
            return bucketIndex;
        }
        int step = 1;
        int currentIndex = this.getNextIndex(bucketIndex, key, step);
        while (currentIndex != bucketIndex ) {
            if ( 
                this.buckets[bucketIndex] != null
                && !this.buckets[bucketIndex].getIsDeleted()  
                && this.buckets[bucketIndex].getKey() == key
            ) {
                return currentIndex;
            }
            currentIndex = this.getNextIndex(currentIndex, key, step++);
        }
        return -1;
    }

    /**
    * O(1)
    * Gets the next index for linear probing.
    */
    private int getNextIndex(int index, K key, int step) {
        return (index+step) % this.capacity;
    }

    /**
    * Average case O(1), Worst case O(n)
    * Finds the next empty index in the hash table.
    */
    private int findEmptyIndex(K key) {
        int index = this.hashFunction(key);
        int step =1;
        while(
            this.buckets[index]!= null
            && !this.buckets[index].getIsDeleted()
        ) {
            index = this.getNextIndex(index, key, step++);
        }
        /**
         *   if (buckets[bucketIndex] == null || buckets[bucketIndex].isIsDeleted())
         */
        return index;
    }
    
}

File: JavaScript/HashMap/CustomLinearProbingHashMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomLinearProbingHashMap = void 0;
var Entry_1 = require("./Entry");
var CustomLinearProbingHashMap = /** @class */ (function () {
    function CustomLinearProbingHashMap(loafFactor, capacity) {
        if (loafFactor === void 0) { loafFactor = 0.75; }
        if (capacity === void 0) { capacity = 8; }
        this.size = 0;
        this.capacity = capacity;
        this.loadFactor = loafFactor;
        this.buckets = new Array(this.capacity);
        this.keys = {};
    }
    /**
     * Checks if the Hash Table is Empty
     * O(1) operation
     * @returns boolean
     */
    CustomLinearProbingHashMap.prototype.isEmpty = function () {
        return this.size === 0;
    };
    /**
     * Gets the next Available index as per Linear Probing
     * O(1) operation
     * @param index
     * @param key
     * @param step
     * @returns number
     */
    CustomLinearProbingHashMap.prototype.getNextIndex = function (index, key, step) {
        return (index + step) % this.capacity;
    };
    /**
    * Loops through the Hash Table and checks if the
    * key is present returns the index returns -1
    * if no key is present.
    * Average case O(1), Worst case O(n)
    * In the average case, it is O(1) due to the uniform distribution of keys.
    * In the worst case, if many collisions occur, it can degrade to O(n).
    * @param key
    * @returns number
    */
    CustomLinearProbingHashMap.prototype.findKey = function (key, callback) {
        var index = this.hashFunction(key);
        if (this.buckets[index] !== undefined
            && this.buckets[index].isDeleted === false
            && this.buckets[index].key === key) {
            return index;
        }
        var step = 1;
        var currentIndex = this.getNextIndex(index, key, step++);
        while (currentIndex != index) {
            if (callback(currentIndex, key, step)) {
                return currentIndex;
            }
            currentIndex = this.getNextIndex(index, key, step++);
        }
        /**
         * It means that the key is not in the hashtable
         * and we return -1
         */
        return -1;
    };
    /**
     * Get the Value for the given Key
     * Average case O(1), Worst case O(n)
     * In the average case, the time complexity is O(1) due to the uniform distribution of keys.
     * In the worst case, if there are many collisions, it can degrade to O(n).
     * @param key
     * @returns V | undefined
     */
    CustomLinearProbingHashMap.prototype.get = function (key) {
        var _this = this;
        var index = this.findKey(key, function (currentIndex, currentKey) {
            return _this.buckets[currentIndex] !== null
                && !_this.buckets[currentIndex].isDeleted
                && _this.buckets[currentIndex].key === currentKey;
        });
        if (index !== -1
            && this.buckets[index] !== null
            && !this.buckets[index].isDeleted) {
            return this.buckets[index].value;
        }
        return undefined;
    };
    /**
     * Adds a new Key Value pair in the Hash Table
     * If the table is nearly full, we resize the HashTable
     * If there is an existing key in the hashtable we update the value
     * Average case O(1), Worst case O(n)
     * In the average case, inserting an element is O(1).
     * In the worst case, if the table is nearly full and many collisions occur, it can degrade to O(n).
     *
     * @param key
     * @param value
     * @returns none
     */
    CustomLinearProbingHashMap.prototype.put = function (key, value) {
        var _this = this;
        var bucketIndex = this.findKey(key, function (currentIndex, currentKey) {
            var _a, _b;
            return (_this.buckets[currentIndex] !== null
                && ((_a = _this.buckets[currentIndex]) === null || _a === void 0 ? void 0 : _a.key) === currentKey
                && !((_b = _this.buckets[currentIndex]) === null || _b === void 0 ? void 0 : _b.isDeleted));
        });
        /**
         *  Don't need to check the buckets[bucketIndex].key === key
         *  as the findKey is basically doing the same
        */
        if (bucketIndex !== -1
            && !this.buckets[bucketIndex].isDeleted) {
            this.buckets[bucketIndex].value = value;
            return;
        }
        var load = (this.size + 1) / this.capacity;
        if (load > this.loadFactor) {
            this.resizeHashTable();
        }
        var newIndex = this.findEmptyIndex(key);
        this.buckets[newIndex] = new Entry_1.Entry(key, value);
        this.keys[key] = newIndex;
        this.size++;
    };
    /**
     *
     * @returns the Set of Keys in the Hash Table
     * O(n)
     */
    CustomLinearProbingHashMap.prototype.keySet = function () {
        return new Set(Object.keys(this.keys));
    };
    /**
     *
     * @returns The Array of Values in the HashTable
     * O(n)
     */
    CustomLinearProbingHashMap.prototype.values = function () {
        return this.buckets.filter(function (entry) { return entry !== null && !entry.isDeleted; }).map(function (entry) { return entry.value; });
    };
    /**
     * Returns the Array of Entries in the HashTable
     * O(n)
     */
    CustomLinearProbingHashMap.prototype.entries = function () {
        return this.buckets.filter(function (entry) { return entry !== null && !entry.isDeleted; });
    };
    /**
     * Removes the Given Key from the Hash Table
     * Average case O(1), Worst case O(n)
     * In the average case, finding and marking an entry as deleted is O(1).
     * In the worst case, if many collisions occur, it can degrade to O(n).
     *
     * @param key
     * @returns Value if thre exists a key or undefined
     */
    CustomLinearProbingHashMap.prototype.remove = function (key) {
        var _this = this;
        var bucketIndex = this.findKey(key, function (currentIndex, currentKey) {
            return (_this.buckets[currentIndex] !== null
                && !_this.buckets[currentIndex].isDeleted
                && _this.buckets[currentIndex].key === currentKey);
        });
        if (bucketIndex !== -1
            && this.buckets[bucketIndex] !== null) {
            this.buckets[bucketIndex].isDeleted = true;
            delete this.keys[key];
            this.size--;
            return this.buckets[bucketIndex].value;
        }
        return undefined;
    };
    /**
     * Checking the presence of a key in the keys object is a constant-time operation.
     * @param key
     * @returns boolean
     */
    CustomLinearProbingHashMap.prototype.constainsKey = function (key) {
        return this.keys.hasOwnProperty(key);
    };
    /**
     * Initializes the buckets array and the keys object.
     * O(n)
     */
    CustomLinearProbingHashMap.prototype.clear = function () {
        this.buckets.fill(new Entry_1.Entry(undefined, undefined));
        this.size = 0;
        this.keys = {};
    };
    /**
     * Rehashes all entries in the hash table.
     * O(n)
     */
    CustomLinearProbingHashMap.prototype.resizeHashTable = function () {
        var _this = this;
        var oldBuckets = this.buckets;
        this.capacity *= 2;
        this.buckets = Array.from({ length: this.capacity }, function () { return new Entry_1.Entry(null, null); });
        this.size = 0;
        this.keys = {};
        oldBuckets.forEach(function (bucketEntry) {
            if (!bucketEntry.isDeleted) {
                _this.put(bucketEntry.key, bucketEntry.value);
            }
        });
    };
    /**
     * Computes the Hash of the key
     * O(|key|) The time complexity depends on the length of the key.
     * @param key
     * @returns
     */
    CustomLinearProbingHashMap.prototype.hashFunction = function (key) {
        // return Number(key) % this.capacity;
        var hash = 0;
        var strKey = String(key);
        for (var i = 0; i < strKey.length; i++) {
            var char = strKey.charCodeAt(i);
            hash = (hash << 5) - hash + char;
            hash |= 0;
        }
        return hash % this.capacity;
    };
    /**
     * Finds the next Empty index in the Hash Table
     * Average case O(1), Worst case O(n)
     * In the average case, it is O(1) due to the uniform distribution of keys.
     * In the worst case, if many collisions occur, it can degrade to O(n).
     * @param key
     * @returns
     */
    CustomLinearProbingHashMap.prototype.findEmptyIndex = function (key) {
        var _a;
        var index = this.hashFunction(key);
        var step = 1;
        while (this.buckets[index] !== null && !((_a = this.buckets[index]) === null || _a === void 0 ? void 0 : _a.isDeleted)) {
            index = this.getNextIndex(index, key, step++);
        }
        return index;
    };
    return CustomLinearProbingHashMap;
}());
exports.CustomLinearProbingHashMap = CustomLinearProbingHashMap;
